pkg/controller/pxc/users.go

```diff
@@ -0,0 +1,280 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const internalPrefix = "internal-"
+
+func (r *ReconcilePerconaXtraDBCluster) reconcileUsers(cr *api.PerconaXtraDBCluster) error {
+	if cr.Status.Status != api.AppStateReady {
+		return nil
+	}
+
+	return r.handleSysUsersSecret(cr)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) handleSysUsersSecret(cr *api.PerconaXtraDBCluster) error {
+	sysUsersSecretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.SecretsName,
+		},
+		&sysUsersSecretObj,
+	)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get sys users secret '%s'", cr.Spec.SecretsName)
+	}
+
+	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
+	if err != nil {
+		return errors.Wrap(err, "marshal sys secret data")
+	}
+	newSecretDataHash := sha256Hash(newSysData)
+
+	internalSysSecretObj, err := r.getInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "get internal sys users secret")
+	}
+
+	dataChanged, err := sysUsersSecretDataChanged(newSecretDataHash, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "check sys users data changes")
+	}
+
+	if !dataChanged {
+		return nil
+	}
+
+	restartPXC, restartProxy, err := r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "manage sys users")
+	}
+
+	err = r.updateInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartProxy && cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Size > 0 {
+		err = r.restartProxy(cr)
+		if err != nil {
+			return errors.Wrap(err, "restart proxy")
+		}
+	}
+
+	if restartPXC {
+		err = r.restartPXC(cr, newSecretDataHash)
+		if err != nil {
+			return errors.Wrap(err, "restart pxc")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBCluster, sysUsersSecretObj, internalSysSecretObj *corev1.Secret) (bool, bool, error) {
+	var restartPXC, restartProxy bool
+	um, err := users.NewManager(cr.Name+"-pxc", string(internalSysSecretObj.Data["root"]))
+	if err != nil {
+		return restartPXC, restartProxy, errors.Wrap(err, "new users manager")
+	}
+
+	var sysUsers []users.SysUser
+
+	for name, pass := range sysUsersSecretObj.Data {
+		hosts := []string{}
+
+		if string(sysUsersSecretObj.Data[name]) == string(internalSysSecretObj.Data[name]) {
+			continue
+		}
+
+		switch name {
+		case "root":
+			restartProxy = true
+			hosts = []string{"localhost", "%"}
+		case "xtrabackup":
+			restartProxy = true
+			restartPXC = true
+			hosts = []string{"localhost"}
+		case "monitor":
+			restartProxy = true
+			hosts = []string{"10.%", "%"}
+		case "clustercheck":
+			restartProxy = true
+			restartPXC = true
```

Comment:
not sure about ProxySQL, is it used?
```suggestion
		case "clustercheck":
			restartPXC = true
```

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/418/commits/6c99594a7fcf4bdb02009113fe96749e10a465f9